### PR TITLE
change: add more prometheus metrics(etcd modify index)

### PR DESCRIPTION
**Pull request by nic-chen**
### What this PR does / why we need it:

add more prometheus metrics for better understanding the situation of APISIX nodes.

#1885 

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/plugins/prometheus/exporter.lua [hunk 1]
```diff
@@ -152,9 +165,70 @@ local function nginx_status()
...
+
+    key_values[1] = key
+    metrics.etcd_modify_indexes:set(max_idx, key_values)
+
+
+    global_max_index = max_idx > global_max_index and max_idx or global_max_index
+
+    return global_max_index
+end
+local function etcd_modify_index()
```
Comment (moonming) [inline apisix/plugins/prometheus/exporter.lua:192]: need test cases for this function

**Comment by membphis** [comment]
Because we added some new metrics this time, it will affect the performance, so let’s do a simple performance test.

**Comment by membphis** [comment]
@nic-chen if you have fixed them, you can click this button, it means you have fixed it.

image

File: apisix/core/config_etcd.lua [hunk 2]
```diff
@@ -221,10 +221,22 @@ local function sync_data(self)
         return true
     end
 
-    local dir_res, err = waitdir(self.etcd_cli, self.key, self.prev_index + 1)
+    local dir_res, err = waitdir(self.etcd_cli, self.key, self.prev_index + 1, self.timeout)
+
     log.info("waitdir key: ", self.key, " prev_index: ", self.prev_index + 1)
     log.info("res: ", json.delay_encode(dir_res, true))
     if not dir_res then
+        -- for fetch the last etcd index
```
Comment (membphis) [inline apisix/core/config_etcd.lua:229]: that is wrong. we need to fetch the last etcd index before `local dir_res, err = waitdir(... ...`

File: apisix/core/config_etcd.lua [hunk 3]
```diff
@@ -221,10 +221,22 @@ local function sync_data(self)
...
+        local key_res, _ = getkey(self.etcd_cli, self.key)
+        if key_res and key_res.headers then
+            local key_index = key_res.headers["X-Etcd-Index"]
+            local key_idx = key_index and tonumber(key_index) or 0
+            if key_idx and key_idx > self.prev_index then
+                -- Avoid the index to exceed 1000 by updating other keys that will causing a full reload
```
Comment (membphis) [inline apisix/core/config_etcd.lua]: this line is too long

Comment (nic-chen) [reply apisix/core/config_etcd.lua]: I have split that into two lines.

File: apisix/core/config_etcd.lua [hunk 4]
```diff
@@ -201,10 +221,24 @@ local function sync_data(self)
         return true
     end
 
-    local dir_res, err = waitdir(self.etcd_cli, self.key, self.prev_index + 1)
+    -- for fetch the etcd index
+    local key_res, _ = getkey(self.etcd_cli, self.key)
```
Comment (membphis) [inline apisix/core/config_etcd.lua:225]: we need to capture the `err` message and check it first

File: apisix/core/config_etcd.lua [hunk 5]
```diff
@@ -228,7 +228,7 @@ local function sync_data(self)
 
     log.info("waitdir key: ", self.key, " prev_index: ", self.prev_index + 1)
     log.info("res: ", json.delay_encode(dir_res, true))
-    if not dir_res then
+    if err and err == "timeout" then
```
Comment (membphis) [inline apisix/core/config_etcd.lua:231]: `if err == "timeout" then` is enough

**Comment by membphis** [comment]
@nic-chen please update the checklist, it seems that we are missing `Have you modified the corresponding document?` .

**Pull request merged**